node/cn/handler_test.go

```diff
@@ -58,3 +63,173 @@ func TestNewProtocolManager(t *testing.T) {
+
+		var pnWithoutBlock []Peer
+		for i := 0; i <= blockReceivingPNLimit; i++ {
+			pnWithoutBlock = append(pnWithoutBlock, mockPN)
+		}
+
+		mockPeers.EXPECT().CNWithoutBlock(hash).Return([]Peer{mockCN}).Times(1)
+		mockPeers.EXPECT().PNWithoutBlock(hash).Return(pnWithoutBlock).Times(1)
+
+		sampledPeers = pm.samplePeersToSendBlock(block)
+		assert.Equal(t, 1+blockReceivingPNLimit, len(sampledPeers))
+		assert.Equal(t, mockCN, sampledPeers[0])
+	}
+
+	// 2. When the given node is a proxy node.
+	{
+		mockPeers := NewMockPeerSet(mockCtrl)
+		pm.nodetype = node.PROXYNODE
+		pm.peers = mockPeers
+
+		mockPeer := NewMockPeer(mockCtrl)
+		mockPeers.EXPECT().PeersWithoutBlockExceptCN(hash).Return([]Peer{mockPeer}).Times(1)
+
+		sampledPeers := pm.samplePeersToSendBlock(block)
+		assert.Equal(t, 1, len(sampledPeers))
+		assert.Equal(t, mockPeer, sampledPeers[0])
+	}
+
+	// 3. When the given node is an end point node.
+	{
+		mockPeers := NewMockPeerSet(mockCtrl)
+		pm.nodetype = node.ENDPOINTNODE
+		pm.peers = mockPeers
+
+		mockPeer := NewMockPeer(mockCtrl)
+		mockPeers.EXPECT().ENWithoutBlock(hash).Return([]Peer{mockPeer}).Times(1)
+
+		sampledPeers := pm.samplePeersToSendBlock(block)
+		assert.Equal(t, 1, len(sampledPeers))
+		assert.Equal(t, mockPeer, sampledPeers[0])
+	}
+
+	// 4. When the given node is a boot node.
+	{
+		pm.nodetype = node.BOOTNODE
+		assert.Equal(t, []Peer{}, pm.samplePeersToSendBlock(block))
+	}
+
+	// 5. When the given node is an unknown node.
+	{
+		pm.nodetype = node.UNKNOWNNODE
+		assert.Equal(t, []Peer{}, pm.samplePeersToSendBlock(block))
+	}
+}
+
+func TestSamplingPeers(t *testing.T) {
+	peers := make([]Peer, 10)
+	assert.Equal(t, peers, samplingPeers(peers, 20))
+	assert.Equal(t, peers[:5], samplingPeers(peers, 5))
+}
+
+func TestBroadcastBlock_NoParentExists(t *testing.T) {
+	pm := &ProtocolManager{}
+	pm.nodetype = node.ENDPOINTNODE
+	block := newBlock(blockNum1)
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	mockBlockChain := mocks.NewMockBlockChain(mockCtrl)
+	mockBlockChain.EXPECT().GetBlock(block.ParentHash(), block.NumberU64()-1).Return(nil).Times(1)
+
+	pm.blockchain = mockBlockChain
+	pm.BroadcastBlock(block)
+}
+
+func TestBroadcastBlock_ParentExists(t *testing.T) {
+	pm := &ProtocolManager{}
+	pm.nodetype = node.ENDPOINTNODE
+	block := newBlock(blockNum1)
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	mockBlockChain := mocks.NewMockBlockChain(mockCtrl)
+	mockBlockChain.EXPECT().GetBlock(block.ParentHash(), block.NumberU64()-1).Return(block).Times(1)
+	mockBlockChain.EXPECT().GetTd(block.ParentHash(), block.NumberU64()-1).Return(big.NewInt(100)).Times(1)
+	pm.blockchain = mockBlockChain
+
+	mockPeers := NewMockPeerSet(mockCtrl)
+	pm.peers = mockPeers
+
+	mockPeer := NewMockPeer(mockCtrl)
+	mockPeers.EXPECT().ENWithoutBlock(block.Hash()).Return([]Peer{mockPeer}).Times(1)
+	mockPeer.EXPECT().AsyncSendNewBlock(block, new(big.Int).Add(block.BlockScore(), big.NewInt(100)))
```

Comment:
How about check the `Times` for this `AsyncSendNewBlock`?

Reply:
Good. I will update the test.